codegen: Resolve overload ambiguities created by defaulted arguments

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49505 Add at::cpu namespace of functions for structured kernels
* #49498 Push anonymous namespace into codegen, not template
* #49490 Make use_c10_dispatcher: full mandatory for structured kernels
* #49489 Move default or no default logic into native.argument
* **#49348 codegen: Resolve overload ambiguities created by defaulted arguments**
* #49341 Tighten up error checking on manual_kernel_registration
* #49245 Construct CppSignatureGroup from NativeFunction

This is a redux of #45666 post refactor, based off of
https://github.com/peterbell10/pytorch/commit/d534f7d4c555a37fd178c143098b8537a5a05d61
Credit goes to peterbell10 for the implementation.

Fixes #43945.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25594004](https://our.internmc.facebook.com/intern/diff/D25594004)

ezyang committed:
codegen: Resolve overload ambiguities created by defaulted arguments

This is a redux of #45666 post refactor, based off of
https://github.com/peterbell10/pytorch/commit/d534f7d4c555a37fd178c143098b8537a5a05d61
Credit goes to peterbell10 for the implementation.

Fixes #43945.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

peterbell10 commented on tools/codegen/api/types.py:
```diff
@@ -145,7 +152,10 @@ def name(self) -> str:
-        cpp_args_str = ', '.join(a.decl() for a in self.arguments())
+        cpp_args_str = ', '.join(
+            a.decl() if a.name not in self.cpp_no_default_args else a.no_default().decl()
+            for a in self.arguments()
```
Curious, why are you keeping the defaults in `cpp.arguments` when they don't exist in the API? 

Also, don't think this interacts with `TensorOptions` at all. So, if that creates ambiguity then there's no way to resolve it.

ezyang replied:
I think you're right, should push this in more. It's kind of annoying if you push this all the way into `cpp.argument` though, since then it becomes pretty wordy.

ezyang committed:
Update on "codegen: Resolve overload ambiguities created by defaulted arguments"


This is a redux of #45666 post refactor, based off of
https://github.com/peterbell10/pytorch/commit/d534f7d4c555a37fd178c143098b8537a5a05d61
Credit goes to peterbell10 for the implementation.

Fixes #43945.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

peterbell10 commented on tools/codegen/api/cpp.py:
```diff
@@ -237,30 +237,40 @@ def default_expr(d: str, t: Type) -> str:
-                default = '{}'
-            elif a.dtype.default == "long":
-                default = 'at::kLong'  # TODO: this is wrong
+            if 'options' not in cpp_no_default_args:
```
The error checking in `NativeFunction.__post_init__` should prevent this from appearing in `cpp_no_default_args`.

ezyang replied:
alright, can turn this into an assert 👍

ezyang committed:
Update on "codegen: Resolve overload ambiguities created by defaulted arguments"


This is a redux of #45666 post refactor, based off of
https://github.com/peterbell10/pytorch/commit/d534f7d4c555a37fd178c143098b8537a5a05d61
Credit goes to peterbell10 for the implementation.

Fixes #43945.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25594004](https://our.internmc.facebook.com/intern/diff/D25594004)

[ghstack-poisoned]

ezyang committed:
Update on "codegen: Resolve overload ambiguities created by defaulted arguments"


This is a redux of #45666 post refactor, based off of
https://github.com/peterbell10/pytorch/commit/d534f7d4c555a37fd178c143098b8537a5a05d61
Credit goes to peterbell10 for the implementation.

Fixes #43945.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25594004](https://our.internmc.facebook.com/intern/diff/D25594004)

[ghstack-poisoned]

ezyang committed:
Update on "codegen: Resolve overload ambiguities created by defaulted arguments"


This is a redux of #45666 post refactor, based off of
https://github.com/peterbell10/pytorch/commit/d534f7d4c555a37fd178c143098b8537a5a05d61
Credit goes to peterbell10 for the implementation.

Fixes #43945.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25594004](https://our.internmc.facebook.com/intern/diff/D25594004)

[ghstack-poisoned]

ezyang committed:
Update on "codegen: Resolve overload ambiguities created by defaulted arguments"


This is a redux of #45666 post refactor, based off of
https://github.com/peterbell10/pytorch/commit/d534f7d4c555a37fd178c143098b8537a5a05d61
Credit goes to peterbell10 for the implementation.

Fixes #43945.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D25594004](https://our.internmc.facebook.com/intern/diff/D25594004)

[ghstack-poisoned]
